apps/documents/serializers.py

```diff
@@ -6,55 +6,105 @@
         max_length=Paragraph._meta.get_field('name').max_length
     )
-    weight = serializers.IntegerField()
-    text = serializers.CharField()
+
+    class Meta:
+        model = Paragraph
+        fields = ('id', 'weight', 'name', 'text')
+        extra_kwargs = {
+            'weight': {'required': False, 'default': 0, 'write_only': True}
+        }
 
 
 class ChapterSerializer(serializers.ModelSerializer):
-    paragraphs = ParagraphSerializer(many=True, partial=True)
+    id = serializers.IntegerField(required=False)
+    paragraphs = ParagraphSerializer(many=True)
 
     class Meta:
         model = Chapter
-        exclude = ('creator', 'module',)
-        extra_kwargs = {'weight': {'required': False}}
-
-    def validate(self, data):
-        module_pk = self._context['module_pk']
-        module = Module.objects.get(pk=module_pk)
-        data['module'] = module
-        return data
+        fields = ('id', 'name', 'weight', 'paragraphs')
+        extra_kwargs = {
+            'weight': {'required': False, 'default': 0, 'write_only': True}
+        }
 
     def create(self, validated_data):
         paragraphs = validated_data.pop('paragraphs')
         user = self.context['request'].user
-        chapter = Chapter.objects.create(creator=user, **validated_data)
+        module = Module.objects.get(pk=self.context['module_pk'])
+        chapter = Chapter.objects.create(creator=user,
+                                         module=module,
+                                         **validated_data)
 
-        for paragraph in paragraphs:
-            Paragraph.objects.create(chapter=chapter, **paragraph)
+        for weight, paragraph_data in enumerate(paragraphs):
+            paragraph_data['weight'] = weight
+            Paragraph.objects.create(chapter=chapter,
+                                     **paragraph_data)
 
         return chapter
 
     def update(self, instance, validated_data):
-        instance.name = validated_data['name']
-        instance.save()
-        paragraphs = validated_data.pop('paragraphs')
+        # Update the chapter
+        paragraphs_data = validated_data.pop('paragraphs')
+        chapter = super(ChapterSerializer, self)\
+            .update(instance, validated_data)
 
-        paragraph_ids = [item['id'] for item in paragraphs if 'id' in item]
-        instance.paragraphs.exclude(id__in=paragraph_ids).delete()
+        # Delete removed paragraphs from the database
+        paragraph_ids = [item['id']
+                         for item in paragraphs_data
+                         if 'id' in item]
+        chapter.paragraphs.exclude(id__in=paragraph_ids).delete()
+
+        # Update or create the chapters
+        for weight, paragraph_data in enumerate(paragraphs_data):
+            paragraph_data['weight'] = weight
+            if 'id' in paragraph_data:
+                paragraph = Paragraph.objects.get(id=paragraph_data['id'])
+                assert paragraph.chapter == chapter
```

Comment:
when validating i don't have the any instances and would have to duplicate the code. it would probably be better, but DRF does not really want you to check on models, but rather the json itself :(